src/proof/lfsc/lfsc_printer.cpp

```diff
@@ -44,7 +44,219 @@ void LfscPrinter::print(std::ostream& out,
+          std::stringstream sscons;
+          sscons << d_tproc.convert(cons.getConstructor());
+          std::string cname =
+              LfscNodeConverter::getNameForUserName(sscons.str());
+          // print construct/tester
+          preamble << "(declare " << cname << " term)" << std::endl;
+          for (size_t j = 0, nargs = cons.getNumArgs(); j < nargs; j++)
+          {
+            const DTypeSelector& arg = cons[j];
+            // print selector
+            Node si = d_tproc.convert(arg.getSelector());
+            std::stringstream sns;
+            sns << si;
+            std::string sname =
+                LfscNodeConverter::getNameForUserName(sns.str());
+            preamble << "(declare " << sname << " term)" << std::endl;
+          }
+        }
+        // testers and updaters are instances of parametric symbols
+        // shared selectors are instance of parametric symbol "sel"
+        preamble << "; END DATATYPE " << std::endl;
+      }
+    }
+  }
+  Trace("lfsc-print-debug") << "; print user symbols" << std::endl;
+  // [1b] user declare function symbols
+  for (const Node& s : syms)
+  {
+    TypeNode st = s.getType();
+    if (st.isConstructor() || st.isSelector() || st.isTester()
+        || st.isUpdater())
+    {
+      // constructors, selector, testers, updaters are defined by the datatype
+      continue;
+    }
+    Node si = d_tproc.convert(s);
+    preamble << "(define " << si << " (var "
+             << d_tproc.getOrAssignIndexForVar(s) << " ";
+    printType(preamble, st);
+    preamble << "))" << std::endl;
+  }
+
+  Trace("lfsc-print-debug") << "; compute proof letification" << std::endl;
+  // [2] compute the proof letification
+  std::vector<const ProofNode*> pletList;
+  std::map<const ProofNode*, size_t> pletMap;
+  computeProofLetification(pnBody, pletList, pletMap);
+
+  Trace("lfsc-print-debug") << "; compute term lets" << std::endl;
+  // compute the term lets
+  LetBinding lbind;
+  for (const Node& ia : iasserts)
+  {
+    lbind.process(ia);
+  }
+  // We do a "dry-run" of proof printing here, using the LetBinding print
+  // channel. This pass traverses the proof but does not print it, but instead
+  // updates the let binding data structure for all nodes that appear anywhere
+  // in the proof.
+  LfscPrintChannelPre lpcp(lbind);
+  LetBinding emptyLetBind;
+  std::map<const ProofNode*, size_t>::iterator itp;
+  for (const ProofNode* p : pletList)
+  {
+    itp = pletMap.find(p);
+    Assert(itp != pletMap.end());
+    size_t pid = itp->second;
+    pletMap.erase(p);
+    printProofInternal(&lpcp, p, emptyLetBind, pletMap, passumeMap);
+    pletMap[p] = pid;
+  }
+  // Print the body of the outermost scope
+  printProofInternal(&lpcp, pnBody, emptyLetBind, pletMap, passumeMap);
+
+  // [3] print warnings
+  for (PfRule r : d_trustWarned)
+  {
+    out << "; WARNING: adding trust step for " << r << std::endl;
+  }
+
+  // [4] print the DSL rewrite rule declarations
```

Comment:
This feels like a TODO, no?
I suggest removing this for now, or adding a TODO (and a corresponding issue).